### PR TITLE
Optimize ADD flag calculation

### PR DESCRIPTION
**Pull request by alyssarosenzweig**
There's An Op For That

### CODE REVIEW

**Comment by alyssarosenzweig** [comment]
Const prop is broken. InstCountCI caught that, nice

**Comment by alyssarosenzweig** [comment]
Segment register code is changing. I don't understand why. @Sonicadvance1 ?

**Comment by Sonicadvance1** [comment]
> Segment register code is changing. I don't understand why. @Sonicadvance1 ?

Oop, need to disable a cmake option `-DENABLE_OFFLINE_TELEMETRY=False`, I should probably add a runtime option so this isn't necessary.

File: FEXCore/Source/Interface/IR/IR.json [hunk 1]
```diff
@@ -895,6 +895,10 @@
           "Size == FEXCore::IR::OpSize::i32Bit || Size == FEXCore::IR::OpSize::i64Bit"
         ]
       },
+      "GPR = AddNZCV u8:$Size, GPR:$Src1, GPR:$Src2": {
```
Comment (Sonicadvance1) [inline FEXCore/Source/Interface/IR/IR.json:898]: This should have an EmitValidation section like the IR op above it so the IR validation can also know that it only supports 32bit and 64bit operation.

Comment (Sonicadvance1) [reply FEXCore/Source/Interface/IR/IR.json:898]: Probably also use `OpSize:#Size` instead of `u8:$Size`

**Comment by alyssarosenzweig** [comment]
Discussion question: would we rather invert CF as stored in the NZCV register? That would mean adding a cfinv to every add, but saving the cfinv from every subtract/compare. The flag consumers the compare could probably tolerate either form. Given that compares are the dominant source of flags, that might come out ahead.

This is the convention that Rosetta uses. That doesn't mean it's the right call but it's not a bad idea either. I don't think this should be addressed in this PR but it's worth further thought for later.

**Comment by Sonicadvance1** [comment]
> Discussion question: would we rather invert CF as stored in the NZCV register? That would mean adding a cfinv to every add, but saving the cfinv from every subtract/compare. The flag consumers the compare could probably tolerate either form. Given that compares are the dominant source of flags, that might come out ahead.
> 
> This is the convention that Rosetta uses. That doesn't mean it's the right call but it's not a bad idea either. I don't think this should be addressed in this PR but it's worth further thought for later.

This will likely need some statistics gathering from random applications to see which would be the bigger win. Hard to know for sure.

**Comment by Sonicadvance1** [comment]
Did some minor testing on this branch and it seems to break something in Steam. Will want to be careful when merging this.

**Comment by alyssarosenzweig** [comment]
Answer:

```c++
  // TODO: Can use OpSize calculated up front.
  //auto OpSize = SrcSize == 8 ? OpSize::i64Bit : OpSize::i32Bit;
  auto OpSize = IR::SizeToOpSize(std::max<uint8_t>(4, std::max(GetOpSize(Src1), GetOpSize(Src2))));
````

Either way should(?) not be a regression from this particular change. Maybe. Although.. hm

**Comment by alyssarosenzweig** [comment]
Seems like it, at the very least, has the potential to screw up the carry/overflow calculations.

**Comment by alyssarosenzweig** [comment]
Pushed an untested patch on top to find out what happens cuz your CI is faster than my Chromebook

File: unittests/InstructionCountCI/PrimaryGroup.json [hunk 2]
```diff
@@ -3375,43 +3287,36 @@
...
       "Optimal": "No",
       "Comment": "GROUP2 0xf7 /2",
       "ExpectedArm64ASM": [
         "mov w20, #0x0",
         "lsr w21, w7, #0",
         "neg w7, w21",
-        "eor x22, x20, x21",
-        "eor x22, x22, x7",
-        "ubfx x22, x22, #4, #1",
+        "eor w22, w20, w21",
```
Comment (alyssarosenzweig) [inline unittests/InstructionCountCI/PrimaryGroup.json:3306]: Note to self: fix const prop to further improve this code?

File: unittests/InstructionCountCI/PrimaryGroup.json [hunk 3]
```diff
@@ -3375,43 +3287,36 @@
...
-        "mrs x22, nzcv",
-        "cmp x20, x21",
-        "cset x23, lo",
-        "orr x22, x22, x23, lsl #29",
-        "eor x21, x20, x21",
-        "eor x20, x7, x20",
-        "and x20, x21, x20",
-        "ubfx x20, x20, #31, #1",
-        "orr x20, x22, x20, lsl #28",
+        "cmp w20, w21",
```
Comment (alyssarosenzweig) [inline unittests/InstructionCountCI/PrimaryGroup.json:3312]: Note to self: extend const prop to handle this so we translate to a `NEGS` instruction instead

**Comment by alyssarosenzweig** [comment]
CI is happy.. @Sonicadvance1 please retest Steam, hopefully that was the problem

**Comment by Sonicadvance1** [comment]
> CI is happy.. @Sonicadvance1 please retest Steam, hopefully that was the problem

Can confirm, Steam is now happy.

**Review by Sonicadvance1 — APPROVED** [review]
Very nice improvements! Nice to see flag generation getting optimized.
Gave this a bunch of testing and I don't see any regressions.

**Pull request merged**